feat(dot/parachain): Implement overseer skeleton

edwardmack opened the pull request:
## Changes
The overseer is responsible for:
1. Setting up, monitoring, and handing failure for overseer subsystems.
2. Providing a "heartbeat" of which relay-parents subsystems should be working on.
3. Acting as a message bus between subsystems.

This PR starts to implement these by implementing how the overseer will start, send OverseerSignal::ActiveLeavesUpdate events to subsystems, and receive messages from subsystems.  This a base to build upon for further implementing functionality needed by the overseer.

<!-- Brief list of functional changes -->

## Tests

<!-- Detail how to run relevant tests to the changes -->
This test shows overseer starting subsystems, and send AciveLeaf and receiving messages.
```sh
go test github.com/ChainSafe/gossamer/dot/parachain/overseer -v
```

## Issues
Issue #3418 
<!-- Write the issue number(s), for example: #123 -->

## Primary Reviewer

<!-- Tag a code owner to review your PR -->
@kishansagathiya 
@timwu20

jimjbrettj commented on dot/parachain/overseer/overseer.go:
```diff
@@ -0,0 +1,74 @@
+	subsystemMessages map[Subsystem]chan any
+}
+
+type ExamlpeSender struct {
```
```suggestion
type ExampleSender struct {
```

jimjbrettj reviewed (approved):
lgtm

kishansagathiya commented on dot/parachain/overseer/overseer.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+func (o *Overseer) RegisterSubsystem(subsystem Subsystem) {
+	// TODO: determine best buffer size
+	comChan := make(chan any, 5)
+	o.subsystemMessages[subsystem] = comChan
+	receiverChan := make(chan any, 5)
```
we should add buffer size on top of the file as constant and use it here.

kishansagathiya commented on dot/parachain/overseer/overseer.go:
```diff
@@ -0,0 +1,74 @@
+			err := sub.Run(ctx)
+			if err != nil {
+				// TODO: handle error
+				fmt.Printf("error running subsystem %v: %v\n", sub, err)
```
We should add a logger on top the file and use that here to log error, sorta like
```
var (
	logger = log.NewFromGlobal(log.AddContext("pkg", "parachain-overseer"))
)
```

kishansagathiya commented on dot/parachain/overseer/overseer.go:
```diff
@@ -0,0 +1,74 @@
+		go func(sub Subsystem, channel chan any) {
+			fmt.Printf("overseer waiting for messages from %v\n", sub)
+
+			for {
```
We should add a return or break in this for loop to end this goroutine. Currently, it is a forever goroutine.

EclesioMeloJunior replied:
Agree, we should pass to the `Start` method a `stop chan struct{}` argument that goroutines should listen for and whenever it triggers all goroutines should return. Also, I would suggest having a `sync.WaitGroup` in the `Overseer` struct and for each goroutine we increment 1 in the wait group

```go
type Overseer struct {
    wg sync.WaitGroup
    doneCh chan struct{}
    stopCh chan struct{}
}

// here stop channel is just a read-only channel which means that
func NewOverseer() *Overseer {
  return &Overseer{
    doneCh: make(chan struct{}),
    stopCh: make(chan struct{}),
  }
}

func (o *Overseer) Start() {
    for subsystem, context := range o.subsystems {
        o.wg.Add(1)
        go func(..., wg *sync.WaitGroup) {
            defer wg.Done()
            select {
            case <-o.stopCh:
               return // here you can shutdown gracefully the goroutine
            case ...
            }
        }
    }

    for subsystem, recChan := range o.subsystemMessages {
        o.wg.Add(1)
        go func(..., wg *sync.WaitGroup) {
            defer wg.Done()
            select {
            case <-o.stopCh:
               return // here you can shutdown gracefully the goroutine
            case ...
            }
        }
    }
}

func (o *Overseer) Stop() {
    if o.doneCh == nil {
        return nil // or return an error 
    }

    close(o.stopCh)
    // once you close the stop channel you should
    // wait for the goroutines to end until
    o.wg.Wait()

    // assign nil to stopCh prevents subsequent calls to
    // Stop method
    o.stopCh = nil  
}
```

kishansagathiya commented on dot/parachain/overseer/overseer.go:
```diff
@@ -0,0 +1,74 @@
+	comChan := make(chan any, 5)
+	o.subsystemMessages[subsystem] = comChan
+	receiverChan := make(chan any, 5)
+	o.subsystems[subsystem] = &Context{
+		Sender:   &ExamlpeSender{senderChan: comChan},
+		Receiver: receiverChan,
+	}
```
I think we would be able to just use a common channel on which all subsystems send and overseer receives. let's call this channel `overseerChan`. We can add this as a field in parachain service struct.

```
    // Declare a channel that can only receive values of type string
    var ch1 <-chan string
    // Declare a channel that can only send values of type string
    var ch2 chan<- string
```
Let's use directional channels to make sure that we can only receive on this channel from overseer and only send from other subsystems. Let's return this `overseerChan` in `NewOverseer`.

For subsystems to receive, we would need different channels for each subsystems. I feel like these channels should be created by different subsystems and passed to overseer on registerSubsystem.

edwardmack replied:
I've updated this to share the overseerChan with subsystems, so the overseer can receive messages from subsystems.  I've added tests to simulate message passing, I think this is working the way that we intend, but I'm not sure I have used directional channels the way we envision.  Let me know your thoughts.

EclesioMeloJunior commented on dot/parachain/overseer/overseer.go:
```diff
@@ -0,0 +1,108 @@
+	"sync"
+)
+
+const COMMS_BUFFER_SIZE = 5
```
Following our project's style (also golang) you could rewrite it using camelCase `CommsBufferSize`

EclesioMeloJunior commented on dot/parachain/overseer/overseer.go:
```diff
@@ -0,0 +1,108 @@
+		o.wg.Add(1)
+		go func(sub Subsystem, ctx *Context, wg sync.WaitGroup) {
+			defer wg.Done()
+			select {
+			case <-o.stopCh:
+				return
+			}
```
you should put `default:` otherwise it will block until `stopCh` is triggered 
```suggestion
			select {
			case <-o.stopCh:
				return
			default:
			}
```

edwardmack replied:
Good point, I also moved this logic to the go routine inside the subsystem, instead of in the go routine that starts the subsystems.

timwu20 commented on dot/parachain/overseer/types.go:
```diff
@@ -0,0 +1,35 @@
+	"github.com/ChainSafe/gossamer/lib/common"
+)
+
+type Context struct {
```
does this need to be public?

edwardmack replied:
no, updated.

timwu20 commented on dot/parachain/overseer/types.go:
```diff
@@ -0,0 +1,35 @@
+	SendMessage(msg any) error
+}
+
+type ActivatedLeaf struct {
+	Hash   common.Hash
+	Number uint32
+}
+
+type ActiveLeavesUpdate struct {
+	Activated *ActivatedLeaf
+}
```
Can you provide comments for exported types? Is this the same terminology they use in polkadot?

edwardmack replied:
Yes, added comments, yes this is the terminology used in polkadot.

timwu20 commented on dot/parachain/overseer/overseer.go:
```diff
@@ -0,0 +1,102 @@
+}
+
+func (o *Overseer) Stop() {
+	if o.doneCh == nil {
+		return
+	}
+	close(o.stopCh)
+	o.wg.Wait()
+	o.stopCh = nil
```
I think you need to add timeout functionality here.  What if the subsystem never stops?

edwardmack replied:
Added.

timwu20 commented on dot/parachain/overseer/overseer.go:
```diff
@@ -0,0 +1,102 @@
+		}(subsystem, context)
+	}
+
+	// wait for messages from subsystems
+	// TODO: this is a temporary solution, we will determine logic to handle different message types
+	for subsystem, recChan := range o.subsystemMessages {
+		go func(sub Subsystem, channel <-chan any) {
+			fmt.Printf("overseer waiting for messages from %v\n", sub)
+			for { //nolint:gosimple
+				select {
+				case msg := <-channel:
+					fmt.Printf("overseer received message from %v: %v\n", sub, msg)
+				}
+			}
+		}(subsystem, recChan)
+	}
+
+	// TODO: add logic to start listening for Block Imported events and Finalisation events
```
I think figuring out these parts may change how the overseer is currently implemented.

edwardmack replied:
I agree, and I think (hope) that the interfaces defined will remain the same, while the functionality is finalized.

timwu20 reviewed (commented):
Take a look at the service architecture discussion https://github.com/ChainSafe/gossamer/discussions/2293.  I feel like this should be implemented using the service architecture, with the ability to send messages.  I do kinda like how the `Sender` is defined as an interface with a synchronous `SendMessage` function that returns an error.  So we don't have to manually determine if a connection is severed by monitoring whether a channel has closed.  But I feel like the `Subsystem` could determine if the connection to the `Overseer` is dropped by closing the `Receiver` channel which would allow you to remove `Context.stopCh`.

edwardmack commented on dot/parachain/overseer/overseer.go:
```diff
@@ -0,0 +1,102 @@
+	overseerChannel   chan any
+}
+
+type ExampleSender struct {
+	senderChan chan any
+}
+
+func (s *ExampleSender) SendMessage(msg any) error {
+	s.senderChan <- msg
+	return nil
+}
```
updated.

edwardmack commented:
I've updated this to use our service architecture, and added wait group timeout.

kanishkatn commented:
I've also come across a usecase while working on disputes-subsystem. In polkadot, they use two sender methods
- `SendMessage()` : Sends a message to the overseer and waits till it gets a response
- `Feed()` : Sends a message to the overseer and exits

I've found this to make the tests easier to write; but not sure how relevant it is for this PR.

edwardmack commented:
Updates: 
- Added use of content.ContextWithCancel to signal subsystems when overseer has canceled their work.
- Use SendMessage interface for subsystems to send messages to overseer.
- Use channels for sending messages from overseer to subsystem.
- Use sync.WaitGroup to monitor when subsystems have stopped.
- added waitTimeout function to end if subsystem waitgroup waits too long.

kishansagathiya commented:
I would like to merge this skeleton PR and do additions and improvements as extra smaller PRs.

EclesioMeloJunior commented on dot/parachain/overseer/overseer.go:
```diff
@@ -0,0 +1,115 @@
+		defer close(c)
+		wg.Wait()
+	}()
+	select {
+	case <-c:
+		return false // completed normally
+	case <-time.After(timeout):
+		return true // timed out
+	}
```
I would suggest using time.NewTimer, since we are avoiding using time.After, using NewTimer you can stop the timer once the channel is closed 
```suggestion
        timeoutTimer := time.NewTimer(timeout)
	select {
	case <-c:
	        if !timeoutTimer.Stop() { <-timeoutTimer.C }
		return false // completed normally
	case <-timeoutTimer.C:
		return true // timed out
	}
```

EclesioMeloJunior commented on dot/parachain/overseer/overseer.go:
```diff
@@ -0,0 +1,115 @@
+	o.subsystems[subsystem] <- update
+}
+
+func waitTimeout(wg *sync.WaitGroup, timeout time.Duration) bool {
```
```suggestion
func waitTimeout(wg *sync.WaitGroup, timeout time.Duration) (timeouted bool) {
```

EclesioMeloJunior reviewed (commented):
LGTM, just a small comment in the `waitTimeout` function

kishansagathiya commented on dot/parachain/overseer/overseer.go:
```diff
@@ -0,0 +1,115 @@
+	for {
+		select {
+		case msg := <-o.SubsystemsToOverseer:
+			switch msg.(type) {
+			default:
+				logger.Error("unknown message type")
+			}
```
This is intentional. Actual message types haven't been added yet. Actual message types could be added along with subsystems. We can just copy message types from polkadot-sdk, but if we do it when we add the subsystem, the types would be more accurate since we would know what we need to send.
If we translate message types from polkadot-sdk, we might end up write a lot of varying data type code, unnecessarily.
